### ISSUE TITLE
Performance issue about the AVRO schema

### ISSUE THREAD

**Issue by RobertIndie** [fsharplang-ru/pulsar-client-dotnet#149]
When I use AVRO schema for message producing and consuming, the performance is much worse than that of JSON schema.

First I performed a single producer and consumer messaging test. The messaging rate of the AVRO schema is much lower than that of JSON, but there is not much difference between the message processing rate of producers and consumers.
Then, I tested it in a relatively complex scenario. Two producers write a pair of messages to two topics, and then one consumer reads data from these two topics, and one producer writes it to the output topic. The messaging rate of the AVRO schema is still much lower than that of JSON, and there is a big difference between the message processing rate of producers and consumers in the AVRO schema test.

Here are my test cases:
* `SimpleJsonSchema`:  A producer sends data to a consumer, using the JSON schema format
* `SimpleAvroSchema`:  A producer sends data to a consumer, using the AVRO schema format
* `TestWithJsonSchema`: Two producers write a pair of messages to two topics using the JSON schema format, and then one consumer reads data from these two topics, and one producer writes it to the output topic.
* `TestWithAvroSchema`: Two producers write a pair of messages to two topics using the AVRO schema format, and then one consumer reads data from these two topics, and one producer writes it to the output topic.

After running each test case for about 20 seconds, I counted the number of input and output messages.
Here are my test results:
* `SimpleJsonSchema`: Input: 2673; Output: 2673
* `SimpleAvroSchema`: Input: 737; Ouput: 722
* `TestWithJsonSchema`: Input: 473; Output: 316
* `TestWithAvroSchema`: Input: 404; Output: 3   <- This is a serious problem.

Here are my testing codes:
```csharp
using System;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Pulsar.Client.Api;
using Pulsar.Client.Common;

namespace Benchmark
{
    class Program
    {
        static async Task Main(string[] args)
        {
            
            await SimpleJsonSchema();
            // await SimpleAvroSchema();
            // await TestWithJsonSchema();
            // await TestWithAvroSchema();
        }
        public class CPMessage
        {
            // Correlation id from the P1 message.
            public int ProducerCorrelationID { get; set; }

            // Correlation id generated for the CP message.
            public int ConsumerProducerCorrelationID { get; set; }

            // The Pulsar generated MessageId of the original P1 message.
            public byte[] ProducerMessageId { get; set; } 


            // Data copied from the original P1 message.
            public byte[] Data { get; set; }
        }
        public class CMessage
        {
            // Correlation id from the P1 message.
            public int ProducerCorrelationID { get; set; }

            // Correlation id generated for the CP1 message.
            public int ConsumerProducerCorrelationID1 { get; set; }

            // Correlation id generated for the CP2 message.
            public int ConsumerProducerCorrelationID2 { get; set; }

            // The Pulsar generated MessageId of the original CP1 message.
            public byte[] ConsumerProducerMessageId1 { get; set; } 

            // The Pulsar generated MessageId of the original CP1 message.
            public byte[] ConsumerProducerMessageId2 { get; set; } 

            // Data copied from the original CP1 message.
            public byte[] CPData1 { get; set; }

            // Data copied from the original CP2 message.
            public byte[] CPData2 { get; set; }
        }

        public class TestMessage
        {
            public int CorrelationID { get; set; }
        }

        static async Task SimpleJsonSchema()
        {
            int produceCount = 0;
            int receiveCount = 0;
            
            var pulsarAddress = "pulsar://localhost:6650";

            String inputTopic = "t1" + Guid.NewGuid().ToString("N");
            String consumerName = "consumer";

            var client = await new PulsarClientBuilder()
                .ServiceUrl(pulsarAddress)
                .BuildAsync();
            CancellationTokenSource cts = new CancellationTokenSource();
            
            var producer1 = await client.NewProducer(Schema.JSON<TestMessage>())
                .Topic(inputTopic)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            await using var subscription1 = await client.NewConsumer(Schema.JSON<TestMessage>())
                .Topic(inputTopic)
                .ConsumerName($"{consumerName}01")
                .SubscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .SubscriptionName($"sub_{consumerName}01")
                .SubscribeAsync();
            var p1t = Task.Run(async () =>
            {
                int i = 0;
                while (!cts.IsCancellationRequested)
                {
                    i++;
                    var message = new TestMessage();
                    message.CorrelationID = i;
                    await producer1.SendAsync(message);
                    Interlocked.Increment(ref produceCount);
                    Console.WriteLine(
                        $"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] P1 Sent message: {i}");
                }
            }, cts.Token);
            var consumeTask = Task.Run(async () =>
            {
                while (!cts.Token.IsCancellationRequested)
                {
                    var message1 = await subscription1.ReceiveAsync();
                    Interlocked.Increment(ref receiveCount);
                    Console.WriteLine($"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] M1: {message1.GetValue().CorrelationID}");
                }
            }, cts.Token);
            Thread.Sleep(20000);
            cts.Cancel();
            Console.WriteLine($"Produce: {produceCount}; Consume: {receiveCount}");
        }
        
        static async Task SimpleAvroSchema()
        {
            int produceCount = 0;
            int receiveCount = 0;
            
            var pulsarAddress = "pulsar://localhost:6650";

            String inputTopic = "t1" + Guid.NewGuid().ToString("N");
            String consumerName = "consumer";

            var client = await new PulsarClientBuilder()
                .ServiceUrl(pulsarAddress)
                .BuildAsync();
            CancellationTokenSource cts = new CancellationTokenSource();
            
            var producer1 = await client.NewProducer(Schema.AVRO<TestMessage>())
                .Topic(inputTopic)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            await using var subscription1 = await client.NewConsumer(Schema.AVRO<TestMessage>())
                .Topic(inputTopic)
                .ConsumerName($"{consumerName}01")
                .SubscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .SubscriptionName($"sub_{consumerName}01")
                .SubscribeAsync();
            var p1t = Task.Run(async () =>
            {
                int i = 0;
                while (!cts.IsCancellationRequested)
                {
                    i++;
                    var message = new TestMessage();
                    message.CorrelationID = i;
                    await producer1.SendAsync(message);
                    Interlocked.Increment(ref produceCount);
                    Console.WriteLine(
                        $"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] P1 Sent message: {i}");
                }
            }, cts.Token);
            var consumeTask = Task.Run(async () =>
            {
                while (!cts.Token.IsCancellationRequested)
                {
                    var message1 = await subscription1.ReceiveAsync();
                    Interlocked.Increment(ref receiveCount);
                    Console.WriteLine($"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] M1: {message1.GetValue().CorrelationID}");
                }
            }, cts.Token);
            Thread.Sleep(20000);
            cts.Cancel();
            Console.WriteLine($"Produce: {produceCount}; Consume: {receiveCount}");
        }
        
        static async Task TestWithJsonSchema()
        {
            int produceCount = 0;
            int consumeCount = 0;
            
            var pulsarAddress = "pulsar://localhost:6650";

            String inputTopic1 = "t1" + Guid.NewGuid().ToString("N");
            String inputTopic2 = "t2" + Guid.NewGuid().ToString("N");
            String outputTopic = "out" + Guid.NewGuid().ToString("N");
            String consumerName = "consumer";

            var client = await new PulsarClientBuilder()
                .ServiceUrl(pulsarAddress)
                .BuildAsync();

            var producer1 = await client.NewProducer(Schema.JSON<CPMessage>())
                .Topic(inputTopic1)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            var producer2 = await client.NewProducer(Schema.JSON<CPMessage>())
                .Topic(inputTopic2)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            var outputProducer = await client.NewProducer(Schema.JSON<CMessage>())
                .Topic(outputTopic)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            CancellationTokenSource cts = new CancellationTokenSource();
            var p1t = Task.Run(async () =>
            {
                int i = 0;
                while (!cts.IsCancellationRequested)
                {
                    i++;
                    var message = new CPMessage();
                    message.ProducerCorrelationID = i;
                    message.ConsumerProducerCorrelationID = i;
                    await producer1.SendAsync(message);
                    Interlocked.Increment(ref produceCount);
                    Console.WriteLine(
                        $"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] P1 Sent message: {i}");
                }
            }, cts.Token);
            var p2t = Task.Run(async () =>
            {
                int i = 0;
                while (!cts.Token.IsCancellationRequested)
                {
                    i++;
                    var message = new CPMessage();
                    message.ProducerCorrelationID = i;
                    message.ConsumerProducerCorrelationID = i;
                    await producer2.SendAsync(message);
                    Interlocked.Increment(ref produceCount);
                    Console.WriteLine(
                        $"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] P2 Sent message: {i}");
                }
            }, cts.Token);

            // Create the first subscription.

            await using var subscription1 = await client.NewConsumer(Schema.JSON<CPMessage>())
                .Topic(inputTopic1)
                .ConsumerName($"{consumerName}01")
                .SubscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .SubscriptionName($"sub_{consumerName}01")
                .SubscribeAsync();

            // Create the second subscription.

            await using var subscription2 = await client.NewConsumer(Schema.JSON<CPMessage>())
                .Topic(inputTopic2)
                .ConsumerName($"{consumerName}02")
                .SubscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .SubscriptionName($"sub_{consumerName}02")
                .SubscribeAsync();
            var mergeTask = Task.Run(async () =>
            {
                while (!cts.Token.IsCancellationRequested)
                {
                    var message1 = await subscription1.ReceiveAsync();
                    var message2 = await subscription2.ReceiveAsync();
                    var cpMsg1 = message1.GetValue();
                    var cpMsg2 = message2.GetValue();
                    var resultingMessage = new CMessage()
                    {
                        ProducerCorrelationID = cpMsg1.ProducerCorrelationID,
                        ConsumerProducerCorrelationID1 = cpMsg1.ConsumerProducerCorrelationID,
                        ConsumerProducerCorrelationID2 = cpMsg2.ConsumerProducerCorrelationID
                    };
                    await subscription1.AcknowledgeAsync(message1.MessageId);
                    await subscription2.AcknowledgeAsync(message2.MessageId);
                    var sendTask = outputProducer.SendAsync(resultingMessage);
                    int taskCount = Task.WaitAny(new Task[]{sendTask}, 100);
                    if (taskCount < 0)
                    {
                        Console.WriteLine("Timeout");
                        await sendTask;
                    }
                    Interlocked.Increment(ref consumeCount);
                    Console.WriteLine($"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] M1: {message1.GetValue().ProducerCorrelationID}; M2: {message2.GetValue().ProducerCorrelationID}");
                }
            }, cts.Token);

            Thread.Sleep(20000);
            cts.Cancel();
            Console.WriteLine($"Produce: {produceCount/2}; Consume: {consumeCount}");
        }
        static async Task TestWithAvroSchema()
        {
            int produceCount = 0;
            int consumeCount = 0;
            
            var pulsarAddress = "pulsar://localhost:6650";

            String inputTopic1 = "t1" + Guid.NewGuid().ToString("N");
            String inputTopic2 = "t2" + Guid.NewGuid().ToString("N");
            String outputTopic = "out" + Guid.NewGuid().ToString("N");
            String consumerName = "consumer";

            var client = await new PulsarClientBuilder()
                .ServiceUrl(pulsarAddress)
                .BuildAsync();

            var producer1 = await client.NewProducer(Schema.AVRO<CPMessage>())
                .Topic(inputTopic1)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            var producer2 = await client.NewProducer(Schema.AVRO<CPMessage>())
                .Topic(inputTopic2)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            var outputProducer = await client.NewProducer(Schema.AVRO<CMessage>())
                .Topic(outputTopic)
                .SendTimeout(TimeSpan.Zero)
                .CreateAsync();
            CancellationTokenSource cts = new CancellationTokenSource();
            var p1t = Task.Run(async () =>
            {
                int i = 0;
                while (!cts.IsCancellationRequested)
                {
                    i++;
                    var message = new CPMessage();
                    message.ProducerCorrelationID = i;
                    message.ConsumerProducerCorrelationID = i;
                    await producer1.SendAsync(message);
                    Interlocked.Increment(ref produceCount);
                    Console.WriteLine(
                        $"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] P1 Sent message: {i}");
                }
            }, cts.Token);
            var p2t = Task.Run(async () =>
            {
                int i = 0;
                while (!cts.Token.IsCancellationRequested)
                {
                    i++;
                    var message = new CPMessage();
                    message.ProducerCorrelationID = i;
                    message.ConsumerProducerCorrelationID = i;
                    await producer2.SendAsync(message);
                    Interlocked.Increment(ref produceCount);
                    Console.WriteLine(
                        $"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] P2 Sent message: {i}");
                }
            }, cts.Token);

            // Create the first subscription.

            await using var subscription1 = await client.NewConsumer(Schema.AVRO<CPMessage>())
                .Topic(inputTopic1)
                .ConsumerName($"{consumerName}01")
                .SubscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .SubscriptionName($"sub_{consumerName}01")
                .SubscribeAsync();

            // Create the second subscription.

            await using var subscription2 = await client.NewConsumer(Schema.AVRO<CPMessage>())
                .Topic(inputTopic2)
                .ConsumerName($"{consumerName}02")
                .SubscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .SubscriptionName($"sub_{consumerName}02")
                .SubscribeAsync();
            var mergeTask = Task.Run(async () =>
            {
                while (!cts.Token.IsCancellationRequested)
                {
                    var message1 = await subscription1.ReceiveAsync();
                    var message2 = await subscription2.ReceiveAsync();
                    var cpMsg1 = message1.GetValue();
                    var cpMsg2 = message2.GetValue();
                    var resultingMessage = new CMessage()
                    {
                        ProducerCorrelationID = cpMsg1.ProducerCorrelationID,
                        ConsumerProducerCorrelationID1 = cpMsg1.ConsumerProducerCorrelationID,
                        ConsumerProducerCorrelationID2 = cpMsg2.ConsumerProducerCorrelationID
                    };
                    await subscription1.AcknowledgeAsync(message1.MessageId);
                    await subscription2.AcknowledgeAsync(message2.MessageId);
                    var sendTask = outputProducer.SendAsync(resultingMessage);
                    int taskCount = Task.WaitAny(new Task[]{sendTask}, 100);
                    if (taskCount < 0)
                    {
                        Console.WriteLine("Timeout");
                        await sendTask;
                    }
                    Interlocked.Increment(ref consumeCount);
                    Console.WriteLine($"[{DateTime.Now:MM/dd/yyy HH:mm:ss.fff}] M1: {message1.GetValue().ProducerCorrelationID}; M2: {message2.GetValue().ProducerCorrelationID}");
                }
            }, cts.Token);

            Thread.Sleep(20000);
            cts.Cancel();
            Console.WriteLine($"Produce: {produceCount/2}; Consume: {consumeCount}");
        }
    }
}
```

Could you help take a look at the reason why the performance of the AVRO schema messaging is worse than that of JSON?
Is there any solution to this problem?

**Comment by Lanayx** [comment]
Thank you, I'll have a look

**Comment by Lanayx** [comment]
@RobertIndie I've just run all tests on my machine and here are the results:

    SimpleJsonSchema: Input: 1267; Output: 1267
    SimpleAvroSchema: Input: 1267; Ouput: 1267
    TestWithJsonSchema: Input: 1283; Output: 1263
    TestWithAvroSchema: Input: 1280; Output: 1258
    
Also adding the results when using sendAndForgetAsync instead of sendAsync

    SimpleJsonSchema: Produce: 163974; Consume: 159431
    SimpleAvroSchema: Produce: 167438; Consume: 132608
    TestWithJsonSchema: Produce: 113807; Consume: 69891
    TestWithAvroSchema: Produce: 126701; Consume: 46019
    
There is still a difference, but not that huge

**Comment by Lanayx** [comment]
I've made additional investigation and found that the performance difference between JSON and Avro holds for POCO-based classes for which Avro schema is generated in runtime. To make Avro deserialization faster it's better to use autogenerated Avro classes (like in [this sample](https://github.com/fsharplang-ru/pulsar-client-dotnet/tree/develop/tests/AvroGenerated)), with this change the performance is very close to JSON.
As a side note - while investigating the issue I've found the memory leak when working with Avro and Json schemas cache and fixed it in 2.0.3 version

**Comment by RobertIndie** [comment]
Thank you! @Lanayx . Version 2.0.3 solved my problem.
